File: apps/ios/Loonext/Features/Settings/ReferralShareBlock.swift
import SwiftUI
import UIKit

/// #288 — one tap, with a message they can edit.
///
/// ## What this had to achieve
///
/// iOS had NO referral surface at all: the mechanism, the accounting and the
/// payout all shipped with #399, and the only client that could reach any of it
/// was the web app. #288 asks for "one tap, a pre-written message they can edit,
/// sent from the phone they are already holding" — and the phone was the one place
/// a contractor could not do it. A crew lead thinking of somebody to tell is
/// standing at a supply counter, not sitting at a laptop.
///
/// ## What it does not do
///
/// The sheet is the system's. The draft goes to the owner's own Messages, WhatsApp
/// or Mail, on their own number, and they pick the recipient — we never see who,
/// and nothing leaves through the carrier. That boundary is why this is not the
/// mass-texting D4 and D11 exclude: the product supplies the words, the person
/// supplies the distribution.
///
/// ## Why the link is not in the text field
///
/// The first owner to rewrite this in their own words would delete it, send it,
/// and get nothing for a referral they actually made. `ReferralShare.shareText`
/// appends it, so no version of this can go out without it.
///
/// Applying: Smart Defaults — the draft is written, because an empty box is a form
/// and #288 is explicit that contractors will not fill one in. Zen of Clarity —
/// one primary action, one fallback, and no formatting controls on a text message.
///
/// PARITY. Word-for-word identical copy to web's `referral-share.tsx` and
/// Android's `ReferralShareBlock.kt`; `ReferralShareTests` asserts it against the
/// shared TypeScript.
struct ReferralShareBlock: View {
    let link: String?
    let code: String

    // AFTER the required `let`s, private and wrapper-initialised, so the
    // memberwise init `ReferralCard` uses is unchanged.
    @Environment(\.appLocale) private var appLocale

    @State private var note = ReferralShare.note
    @State private var copied = false

    private var text: String {
        ReferralShare.shareText(note: note, link: link, code: code, locale: appLocale)
    }

    private func t(_ key: String) -> String {
        AppStrings.translate(appLocale, key)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(t(ReferralShare.draftLabelKey))
                .font(.golos(11.5, weight: .medium))
                .foregroundStyle(BrandColor.muted600)

            TextField(t(ReferralShare.noteKey), text: $note, axis: .vertical)
                .font(.golos(13))
                .textFieldStyle(.roundedBorder)
                .lineLimit(3 ... 8)

            // Concatenated rather than interpolated into a string literal: the
            // note is `ReferralShare`'s, asserted against the shared TypeScript,
            // and a literal wrapped around it reads to the #228 scanner as a
            // sentence typed in here that no translator will ever see.
            Text(t(ReferralShare.linkNoteKey) + " " + (link ?? code))
                .font(.golos(11.5))
                .foregroundStyle(BrandColor.muted600)

            HStack(spacing: 8) {
                ShareLink(item: text) {
                    Label(t(ReferralShare.actionKey), systemImage: "square.and.arrow.up")
                        .font(.golos(13, weight: .medium))
                }
                .buttonStyle(.borderedProminent)

                Button {
                    UIPasteboard.general.string = text
                    copied = true
                } label: {
                    Text(t(copied ? ReferralShare.copiedKey : ReferralShare.copyKey))
                        .font(.golos(13, weight: .medium))
                }
                .buttonStyle(.bordered)
            }
        }
        // #228: the draft is the one string on this card that GOES OUT, and a
        // French owner sending an English pitch is the whole defect. It cannot
        // be set at init — `@State`'s initial value is computed before this view
        // has an environment to read — so the reader's copy is handed over on
        // appear instead.
        //
        // GUARDED ON THE ENGLISH DEFAULT, which is what makes it safe to run
        // every time: once somebody has typed anything, `note` no longer equals
        // it and their words are never replaced. A reader whose language is
        // English matches on both sides and nothing changes.
        .onAppear {
            if note == ReferralShare.note {
                note = ReferralShare.localisedNote(appLocale)
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/ReferralShareBlock.kt
package com.loonext.android.features.settings

import android.content.Context
import android.content.Intent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.unit.dp
import com.loonext.android.core.referral.ReferralShare

/**
 * #288 — one tap, with a message they can edit.
 *
 * ## What this had to achieve
 *
 * Android had NO referral surface at all: the link existed on the server and on
 * the web app, and the phone — which is the device a contractor actually has in
 * their hand when they think of somebody to tell — could not reach it. #288 asks
 * for "one tap, a pre-written message they can edit, sent from the phone they are
 * already holding", and this is the phone half of that.
 *
 * ## What it does not do
 *
 * The chooser is Android's. The draft goes to the owner's own Messages, WhatsApp
 * or email, on their own number, and they pick the recipient — we never see who,
 * and nothing leaves through the carrier. That boundary is why this is not the
 * mass-texting D4 and D11 exclude: the product supplies the words, the person
 * supplies the distribution.
 *
 * ## Why the link is not in the text field
 *
 * The first owner to rewrite this in their own words would delete it, send it, and
 * get nothing for a referral they actually made. [ReferralShare.shareText] appends
 * it, so no version of this can go out without it.
 *
 * Applying: Smart Defaults — the draft is written, because an empty box is a form
 * and #288 is explicit that contractors will not fill one in. Zen of Clarity — one
 * primary action, one fallback, and no formatting controls on a text message.
 *
 * PARITY. Word-for-word identical copy to web's `referral-share.tsx` and iOS's
 * `ReferralShareBlock.swift`; `ReferralShareTest` asserts it against the shared
 * TypeScript.
 */
@Composable
fun ReferralShareBlock(link: String?, code: String) {
    var note by remember { mutableStateOf(ReferralShare.NOTE) }
    var copied by remember { mutableStateOf(false) }
    val context = LocalContext.current
    val clipboard = LocalClipboardManager.current
    val text = ReferralShare.shareText(note, link, code)
    // #228: the words are `ReferralShare`'s and stay there — they are asserted
    // against the shared TypeScript by `ReferralShareTest`, so a copy in the
    // Android catalogue would be a second source for one sentence. Built here
    // rather than inline so what reaches `Text` is a value, not a template.
    val linkLine = "${ReferralShare.LINK_NOTE} ${link ?: code}"

    OutlinedTextField(
        value = note,
        onValueChange = { note = it },
        label = { Text(ReferralShare.DRAFT_LABEL) },
        minLines = 3,
        modifier = Modifier.fillMaxWidth(),
    )
    Spacer(Modifier.height(6.dp))
    Text(
        linkLine,
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(Modifier.height(10.dp))
    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Button(onClick = { shareReferral(context, text) }) {
            Text(ReferralShare.ACTION)
        }
        OutlinedButton(
            onClick = {
                clipboard.setText(AnnotatedString(text))
                copied = true
            },
        ) {
            Text(if (copied) ReferralShare.COPIED else ReferralShare.COPY)
        }
    }
}

/**
 * The chooser. Same shape as the diagnostics and crash-report shares.
 *
 * `runCatching` because a device with no app that accepts text/plain would
 * otherwise crash the settings screen — a stripped fleet tablet is rare and a
 * crash is not the honest response to it.
 */
internal fun shareReferral(context: Context, text: String) {
    val send = Intent(Intent.ACTION_SEND)
        .setType("text/plain")
        .putExtra(Intent.EXTRA_TEXT, text)
    runCatching {
        context.startActivity(Intent.createChooser(send, ReferralShare.TITLE))
    }
}
